File: html_to_bark.py
#!/usr/bin/env python3

from __future__ import annotations

import html
import os
import re
import sys
from dataclasses import dataclass, field
from html.parser import HTMLParser


VOID_TAGS = {
    "area",
    "base",
    "br",
    "col",
    "embed",
    "hr",
    "img",
    "input",
    "link",
    "meta",
    "param",
    "source",
    "track",
    "wbr",
}


@dataclass
class TextNode:
    text: str


@dataclass
class ElementNode:
    tag: str
    attrs: list[tuple[str, str | None]] = field(default_factory=list)
    children: list[ElementNode | TextNode] = field(default_factory=list)


class TreeBuilder(HTMLParser):
    def __init__(self) -> None:
        super().__init__(convert_charrefs=True)
        self.root = ElementNode("root")
        self.stack = [self.root]

    def handle_starttag(self, tag: str, attrs: list[tuple[str, str | None]]) -> None:
        node = ElementNode(tag, attrs)
        self.stack[-1].children.append(node)
        if tag not in VOID_TAGS:
            self.stack.append(node)

    def handle_endtag(self, tag: str) -> None:
        for i in range(len(self.stack) - 1, 0, -1):
            if self.stack[i].tag == tag:
                del self.stack[i:]
                break

    def handle_startendtag(self, tag: str, attrs: list[tuple[str, str | None]]) -> None:
        self.stack[-1].children.append(ElementNode(tag, attrs))

    def handle_data(self, data: str) -> None:
        if not data:
            return
        self.stack[-1].children.append(TextNode(data))

    def handle_entityref(self, name: str) -> None:
        self.handle_data(html.unescape(f"&{name};"))

    def handle_charref(self, name: str) -> None:
        self.handle_data(html.unescape(f"&#{name};"))


def normalize_text(text: str) -> str:
    return re.sub(r"\s+", " ", text).strip()


def is_significant_text(node: TextNode) -> bool:
    return normalize_text(node.text) != ""


def format_attr_value(value: str) -> str:
    if re.fullmatch(r'[^=\s\[\]"]+', value):
        return value
    escaped = value.replace('"', '\\"')
    return f'"{escaped}"'


def split_ids(value: str) -> list[str]:
    return [part for part in value.split() if part]


def split_classes(value: str) -> list[str]:
    return [part for part in value.split() if part]


def format_node(node: ElementNode | TextNode, indent: int = 0) -> list[str]:
    if isinstance(node, TextNode):
        text = normalize_text(node.text)
        return [(" " * indent) + text] if text else []

    indent_str = " " * indent
    tag = "" if node.tag == "div" else node.tag

    ids: list[str] = []
    classes: list[str] = []
    other_attrs: list[tuple[str, str | None]] = []
    for key, value in node.attrs:
        value = "" if value is None else value
        if key == "id":
            ids = split_ids(value)
        elif key == "class":
            classes = split_classes(value)
        else:
            other_attrs.append((key, value))

    head = "["
    if tag:
        head += tag

    if ids:
        if head == "[":
            head += f"@{ids[0]}"
            ids = ids[1:]
        for item in ids:
            head += f" @{item}"

    if classes:
        if head == "[":
            head += "<: "
        else:
            head += " <: "
        head += ", ".join(classes)

    if other_attrs:
        attrs_str = " ".join(f"{key}={format_attr_value(value)}" for key, value in other_attrs)
        if head == "[":
            head += attrs_str
        else:
            head += " " + attrs_str
    has_metadata = bool(ids or classes or other_attrs)

    children = [child for child in node.children if not isinstance(child, TextNode) or is_significant_text(child)]
    if not children:
        return [indent_str + head + "]"]

    only_text = all(isinstance(child, TextNode) for child in children)
    if only_text:
        body = " ".join(normalize_text(child.text) for child in children if isinstance(child, TextNode))
        if has_metadata:
            return [f"{indent_str}{head} | {body}]"]
        if tag:
            return [f"{indent_str}{head} {body}]"]
        return [f"{indent_str}{head}{body}]"]

    lines = [indent_str + head]
    for child in children:
        lines.extend(format_node(child, indent + 2))
    lines.append(indent_str + "]")
    return lines


def convert_html_to_bark(source: str) -> str:
    parser = TreeBuilder()
    parser.feed(source)
    lines: list[str] = []
    for child in parser.root.children:
        lines.extend(format_node(child, 0))
    return "\n".join(lines) + "\n"


def main() -> int:
    if len(sys.argv) != 2:
        print(f"usage: {os.path.basename(sys.argv[0])} <source-html>", file=sys.stderr)
        return 2

    with open(sys.argv[1], "r", encoding="utf-8") as fh:
        source = fh.read()

    sys.stdout.write(convert_html_to_bark(source))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: htmltobark.go
package main

import (
	"fmt"
	"io"
	"os"
	"regexp"
	"strings"
	"unicode"
)

var whitespaceRE = regexp.MustCompile(`\s+`)

var voidTags = map[string]bool{
	"area":   true,
	"base":   true,
	"br":     true,
	"col":    true,
	"embed":  true,
	"hr":     true,
	"img":    true,
	"input":  true,
	"link":   true,
	"meta":   true,
	"param":  true,
	"source": true,
	"track":  true,
	"wbr":    true,
}

type node interface {
	isNode()
}

type textNode struct {
	Text string
}

func (textNode) isNode() {}

type elementNode struct {
	Tag      string
	Attrs    [][2]string
	Children []node
}

func (elementNode) isNode() {}

type parser struct {
	src string
	pos int
}

func main() {
	if len(os.Args) != 2 {
		fmt.Fprintf(os.Stderr, "usage: %s <source-html>\n", os.Args[0])
		os.Exit(2)
	}

	input, err := os.ReadFile(os.Args[1])
	if err != nil {
		fmt.Fprintf(os.Stderr, "read source: %v\n", err)
		os.Exit(1)
	}

	out, err := ConvertHTMLToBark(string(input))
	if err != nil {
		fmt.Fprintf(os.Stderr, "convert html to bark: %v\n", err)
		os.Exit(1)
	}

	if _, err := io.WriteString(os.Stdout, out); err != nil {
		fmt.Fprintf(os.Stderr, "write output: %v\n", err)
		os.Exit(1)
	}
}

func ConvertHTMLToBark(source string) (string, error) {
	p := &parser{src: source}
	nodes, err := p.parseNodes("")
	if err != nil {
		return "", err
	}

	var lines []string
	for _, n := range nodes {
		lines = append(lines, formatNode(n, 0)...)
	}
	return strings.Join(lines, "\n") + "\n", nil
}

func (p *parser) parseNodes(stopTag string) ([]node, error) {
	var nodes []node
	for !p.eof() {
		if p.startsWith("<!--") {
			p.skipComment()
			continue
		}
		if p.startsWith("<!") {
			p.skipDeclaration()
			continue
		}
		if stopTag != "" && p.startsWith("</") {
			tag, err := p.parseClosingTag()
			if err != nil {
				return nil, err
			}
			if strings.EqualFold(tag, stopTag) {
				return nodes, nil
			}
			continue
		}
		if p.startsWith("<") {
			elem, selfClosing, err := p.parseStartTag()
			if err != nil {
				return nil, err
			}
			if !selfClosing && !voidTags[strings.ToLower(elem.Tag)] {
				if strings.EqualFold(elem.Tag, "script") || strings.EqualFold(elem.Tag, "style") {
					raw, err := p.readRawUntilClosingTag(elem.Tag)
					if err != nil {
						return nil, err
					}
					if raw != "" {
						elem.Children = append(elem.Children, textNode{Text: unescapeHTML(raw)})
					}
				} else {
					children, err := p.parseNodes(elem.Tag)
					if err != nil {
						return nil, err
					}
					elem.Children = children
				}
			}
			nodes = append(nodes, elem)
			continue
		}

		text := p.readUntil("<")
		if text != "" {
			nodes = append(nodes, textNode{Text: unescapeHTML(text)})
		}
	}

	if stopTag != "" {
		return nil, fmt.Errorf("unterminated <%s>", stopTag)
	}
	return nodes, nil
}

func (p *parser) parseStartTag() (elementNode, bool, error) {
	if !p.consume("<") {
		return elementNode{}, false, fmt.Errorf("expected '<' at %d", p.pos)
	}

	tag := p.readName()
	if tag == "" {
		return elementNode{}, false, fmt.Errorf("expected tag name at %d", p.pos)
	}

	var attrs [][2]string
	selfClosing := false

	for !p.eof() {
		p.skipSpaces()
		switch {
		case p.startsWith("/>"):
			p.pos += 2
			selfClosing = true
			return elementNode{Tag: tag, Attrs: attrs}, selfClosing, nil
		case p.startsWith(">"):
			p.pos++
			return elementNode{Tag: tag, Attrs: attrs}, selfClosing, nil
		default:
			key := p.readAttrName()
			if key == "" {
				return elementNode{}, false, fmt.Errorf("expected attribute name in <%s> at %d", tag, p.pos)
			}
			p.skipSpaces()
			value := ""
			if p.consume("=") {
				p.skipSpaces()
				value = p.readAttrValue()
			}
			attrs = append(attrs, [2]string{key, unescapeHTML(value)})
		}
	}

	return elementNode{}, false, fmt.Errorf("unterminated <%s>", tag)
}

func (p *parser) parseClosingTag() (string, error) {
	if !p.consume("</") {
		return "", fmt.Errorf("expected closing tag at %d", p.pos)
	}
	tag := p.readName()
	p.skipSpaces()
	if !p.consume(">") {
		return "", fmt.Errorf("unterminated closing tag </%s>", tag)
	}
	return tag, nil
}

func (p *parser) readRawUntilClosingTag(tag string) (string, error) {
	lower := strings.ToLower(p.src[p.pos:])
	needle := "</" + strings.ToLower(tag) + ">"
	idx := strings.Index(lower, needle)
	if idx < 0 {
		return "", fmt.Errorf("unterminated raw <%s> block", tag)
	}
	raw := p.src[p.pos : p.pos+idx]
	p.pos += idx + len(needle)
	return raw, nil
}

func (p *parser) skipComment() {
	if idx := strings.Index(p.src[p.pos:], "-->"); idx >= 0 {
		p.pos += idx + 3
		return
	}
	p.pos = len(p.src)
}

func (p *parser) skipDeclaration() {
	if idx := strings.Index(p.src[p.pos:], ">"); idx >= 0 {
		p.pos += idx + 1
		return
	}
	p.pos = len(p.src)
}

func (p *parser) readUntil(token string) string {
	if idx := strings.Index(p.src[p.pos:], token); idx >= 0 {
		out := p.src[p.pos : p.pos+idx]
		p.pos += idx
		return out
	}
	out := p.src[p.pos:]
	p.pos = len(p.src)
	return out
}

func (p *parser) readName() string {
	start := p.pos
	for !p.eof() {
		r := rune(p.src[p.pos])
		if unicode.IsLetter(r) || unicode.IsDigit(r) || r == '-' || r == '_' {
			p.pos++
			continue
		}
		break
	}
	return p.src[start:p.pos]
}

func (p *parser) readAttrName() string {
	start := p.pos
	for !p.eof() {
		r := rune(p.src[p.pos])
		if unicode.IsLetter(r) || unicode.IsDigit(r) || r == '-' || r == '_' || r == ':' {
			p.pos++
			continue
		}
		break
	}
	return p.src[start:p.pos]
}

func (p *parser) readAttrValue() string {
	if p.eof() {
		return ""
	}
	switch p.src[p.pos] {
	case '"', '\'':
		quote := p.src[p.pos]
		p.pos++
		start := p.pos
		for !p.eof() && p.src[p.pos] != quote {
			p.pos++
		}
		value := p.src[start:p.pos]
		if !p.eof() {
			p.pos++
		}
		return value
	default:
		start := p.pos
		for !p.eof() {
			r := rune(p.src[p.pos])
			if unicode.IsSpace(r) || r == '>' || (r == '/' && p.pos+1 < len(p.src) && p.src[p.pos+1] == '>') {
				break
			}
			p.pos++
		}
		return p.src[start:p.pos]
	}
}

func (p *parser) skipSpaces() {
	for !p.eof() && unicode.IsSpace(rune(p.src[p.pos])) {
		p.pos++
	}
}

func (p *parser) startsWith(prefix string) bool {
	return strings.HasPrefix(p.src[p.pos:], prefix)
}

func (p *parser) consume(prefix string) bool {
	if p.startsWith(prefix) {
		p.pos += len(prefix)
		return true
	}
	return false
}

func (p *parser) eof() bool {
	return p.pos >= len(p.src)
}

func normalizeText(text string) string {
	return strings.TrimSpace(whitespaceRE.ReplaceAllString(text, " "))
}

func formatAttrValue(value string) string {
	if value != "" && regexp.MustCompile(`^[^=\s\[\]"]+$`).MatchString(value) {
		return value
	}
	return `"` + strings.ReplaceAll(value, `"`, `\"`) + `"`
}

func splitFields(value string) []string {
	return strings.Fields(value)
}

func formatNode(n node, indent int) []string {
	switch typed := n.(type) {
	case textNode:
		text := normalizeText(typed.Text)
		if text == "" {
			return nil
		}
		return []string{strings.Repeat(" ", indent) + text}
	case elementNode:
		return formatElement(typed, indent)
	default:
		return nil
	}
}

func formatElement(elem elementNode, indent int) []string {
	indentStr := strings.Repeat(" ", indent)
	tag := elem.Tag
	if tag == "div" {
		tag = ""
	}

	var ids []string
	var classes []string
	var otherAttrs [][2]string
	for _, attr := range elem.Attrs {
		switch attr[0] {
		case "id":
			ids = splitFields(attr[1])
		case "class":
			classes = splitFields(attr[1])
		default:
			otherAttrs = append(otherAttrs, attr)
		}
	}

	head := "["
	if tag != "" {
		head += tag
	}
	if len(ids) > 0 {
		if head == "[" {
			head += "@" + ids[0]
			ids = ids[1:]
		}
		for _, id := range ids {
			head += " @" + id
		}
	}
	if len(classes) > 0 {
		if head == "[" {
			head += "<: "
		} else {
			head += " <: "
		}
		head += strings.Join(classes, ", ")
	}
	if len(otherAttrs) > 0 {
		parts := make([]string, 0, len(otherAttrs))
		for _, attr := range otherAttrs {
			parts = append(parts, attr[0]+"="+formatAttrValue(attr[1]))
		}
		if head == "[" {
			head += strings.Join(parts, " ")
		} else {
			head += " " + strings.Join(parts, " ")
		}
	}
	hasMetadata := len(ids) > 0 || len(classes) > 0 || len(otherAttrs) > 0

	var children []node
	for _, child := range elem.Children {
		switch c := child.(type) {
		case textNode:
			if normalizeText(c.Text) == "" {
				continue
			}
		}
		children = append(children, child)
	}

	if len(children) == 0 {
		return []string{indentStr + head + "]"}
	}

	onlyText := true
	var textParts []string
	for _, child := range children {
		t, ok := child.(textNode)
		if !ok {
			onlyText = false
			break
		}
		textParts = append(textParts, normalizeText(t.Text))
	}
	if onlyText {
		body := strings.Join(textParts, " ")
		if hasMetadata {
			return []string{fmt.Sprintf("%s%s | %s]", indentStr, head, body)}
		}
		if tag != "" {
			return []string{fmt.Sprintf("%s%s %s]", indentStr, head, body)}
		}
		return []string{fmt.Sprintf("%s%s%s]", indentStr, head, body)}
	}

	lines := []string{indentStr + head}
	for _, child := range children {
		lines = append(lines, formatNode(child, indent+2)...)
	}
	lines = append(lines, indentStr+"]")
	return lines
}

func unescapeHTML(s string) string {
	replacer := strings.NewReplacer(
		"&amp;", "&",
		"&lt;", "<",
		"&gt;", ">",
		"&quot;", `"`,
		"&#39;", "'",
		"&nbsp;", " ",
	)
	return replacer.Replace(s)
}
